### PR TITLE
Change RestGuildMemberAddPayload to use the ids of the roles...

### PR DESCRIPTION
**Pull request by Plerx2493**
...which should be added. Also updates the associated methods

### CODE REVIEW

File: DSharpPlus/Entities/Guild/DiscordGuild.cs [hunk 1]
```diff
@@ -921,6 +921,55 @@ public async Task<IReadOnlyList<DiscordMember>> SearchMembersAsync(string name,
...
+    /// <exception cref="NotFoundException">Thrown when the <paramref name="user"/> or <paramref name="accessToken"/> is not found.</exception>
+    /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+    /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+    public async Task<DiscordMember?> AddMemberAsync
+    (
+        DiscordUser user,
+        string accessToken,
+        string? nickname = null,
+        IEnumerable<ulong>? roles = null,
+        bool muted = false, bool deaf = false
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Guild/DiscordGuild.cs:947]: nit: formatting

**Review by akiraveliara — APPROVED** [review]

File: DSharpPlus/Entities/Guild/DiscordGuild.cs (hunk 1 above)
Comment (Plerx2493) [reply DSharpPlus/Entities/Guild/DiscordGuild.cs]: should be resolved

File: DSharpPlus/Entities/Guild/DiscordGuild.cs [hunk 2]
```diff
@@ -918,12 +916,111 @@ public async Task<IReadOnlyList<DiscordMember>> SearchMembersAsync(string name,
...
+    /// <returns>Only returns the member if they were not already in the guild</returns>
+    /// <exception cref="UnauthorizedException">Thrown when the client does not have the <see cref="DiscordPermissions.CreateInstantInvite" /> permission.</exception>
+    /// <exception cref="NotFoundException">Thrown when the <paramref name="user"/> or <paramref name="accessToken"/> is not found.</exception>
+    /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+    /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+    public async Task<DiscordMember?> AddMemberWithRolesAsync
+    (
+        DiscordUser user,
+        string accessToken,
+        IEnumerable<ulong>? roles,
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Guild/DiscordGuild.cs:942]: why would this be nullable?

File: DSharpPlus/Entities/Guild/DiscordGuild.cs [hunk 3]
```diff
@@ -918,12 +916,111 @@ public async Task<IReadOnlyList<DiscordMember>> SearchMembersAsync(string name,
...
+    /// <returns>Only returns the member if they were not already in the guild</returns>
+    /// <exception cref="UnauthorizedException">Thrown when the client does not have the <see cref="DiscordPermissions.CreateInstantInvite" /> permission.</exception>
+    /// <exception cref="NotFoundException">Thrown when the <paramref name="userId"/> or <paramref name="accessToken"/> is not found.</exception>
+    /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+    /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+    public async Task<DiscordMember?> AddMemberWithRolesAsync
+    (
+        ulong userId,
+        string accessToken,
+        IEnumerable<ulong>? roles,
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Guild/DiscordGuild.cs:967]: here too

File: DSharpPlus/Entities/Guild/DiscordGuild.cs [hunk 4]
```diff
@@ -918,12 +916,111 @@ public async Task<IReadOnlyList<DiscordMember>> SearchMembersAsync(string name,
...
+    /// <returns>Only returns the member if they were not already in the guild</returns>
+    /// <exception cref="UnauthorizedException">Thrown when the client does not have the <see cref="DiscordPermissions.CreateInstantInvite" /> permission.</exception>
+    /// <exception cref="NotFoundException">Thrown when the <paramref name="user"/> or <paramref name="accessToken"/> is not found.</exception>
+    /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+    /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+    public async Task<DiscordMember?> AddMemberWithRolesAsync
+    (
+        DiscordUser user,
+        string accessToken,
+        IEnumerable<DiscordRole>? roles,
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Guild/DiscordGuild.cs:992]: here too

File: DSharpPlus/Entities/Guild/DiscordGuild.cs [hunk 5]
```diff
@@ -918,12 +916,111 @@ public async Task<IReadOnlyList<DiscordMember>> SearchMembersAsync(string name,
...
+    /// <returns>Only returns the member if they were not already in the guild</returns>
+    /// <exception cref="UnauthorizedException">Thrown when the client does not have the <see cref="DiscordPermissions.CreateInstantInvite" /> permission.</exception>
+    /// <exception cref="NotFoundException">Thrown when the <paramref name="userId"/> or <paramref name="accessToken"/> is not found.</exception>
+    /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+    /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+    public async Task<DiscordMember?> AddMemberWithRolesAsync
+    (
+        ulong userId,
+        string accessToken,
+        IEnumerable<DiscordRole>? roles,
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Guild/DiscordGuild.cs:1017]: here too

**Review by akiraveliara — APPROVED** [review]

**Comment by akiraveliara** [comment]
add to #1585

**Pull request merged**